hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java

```diff
@@ -0,0 +1,252 @@
+                    return;
+                }
+            }
+
+            // Try to run this step in this thread, otherwise
+            // offload the step to relevant executor(it
+            // is operation or general-purpose executor)
+            if (!runDirect(step)) {
+                offloadRun(step, this);
+                return;
+            }
+
+            // if there are many offloadedOperations belonging for
+            // a specific map, this step-runner tries to run all in
+            // one go. Let's assume all steps are partition-runnable,
+            // and can run on partition thread, in this case, to
+            // prevent unfair usage of partition thread, we put
+            // an upper time-limit with maxRunNanos, so a map
+            // cannot make a partition thread busy indefinitely
+            // and other operations can have a chance to run.
+            if (maxRunNanos > 0 && runningOnPartitionThread
+                    && System.nanoTime() - start >= maxRunNanos) {
+                step = stepSupplier.get();
+                if (step != null) {
+                    offloadRun(step, this);
+                    return;
+                }
+            }
+        } while (true);
+    }
+
+    /**
+     * @return null if no offloaded operations to execute, otherwise
+     * create next step supplier for the next offloaded operation
+     */
+    @Nullable
+    private StepSupplier getNextStepSupplierOrNull() {
+        for (MapOperation operation : offloadedOperations) {
+            return new StepSupplier(operation);
+        }
+
+        return null;
+    }
+
+    private boolean runDirect(Runnable step) {
+        if (step instanceof PartitionSpecificRunnable) {
+            if (isRunningOnPartitionThread()) {
+                try {
+                    CURRENTLY_EXECUTING_ON_PARTITION_THREAD.set(true);
+                    step.run();
+                } finally {
+                    CURRENTLY_EXECUTING_ON_PARTITION_THREAD.set(false);
+                }
+                return true;
+            }
+        } else {
+            if (!isRunningOnPartitionThread()) {
+                step.run();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void offloadRun(Runnable step,
+                            PartitionSpecificRunnable offload) {
+        if (step instanceof PartitionSpecificRunnable) {
+            operationExecutor.execute(offload);
+        } else {
+            executor.execute(offload);
+        }
+    }
+
+    private Set<MapOperation> getOffloadedOperations(MapOperation mapOperation) {
+        return mapOperation.getRecordStore().getOffloadedOperations();
+    }
+
+    @Override
+    public int getPartitionId() {
+        return partitionId;
+    }
+
+    /**
+     * Removes response sent operation from offloadedOperations.
+     */
+    private class StepResponseHandler implements OperationResponseHandler {
```

Comment:
This name can be confused with the other `StepResponseHandler`. `OffloadedStepResponseHandler` would be better.

Reply:
done.